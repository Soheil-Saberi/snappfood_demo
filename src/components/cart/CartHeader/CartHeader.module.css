.div{
    @apply flex flex-row justify-between
}

.div img{
    @apply h-5
}