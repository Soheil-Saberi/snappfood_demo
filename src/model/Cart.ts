export default interface Cart {
  productId: number;
  count: number;
}
