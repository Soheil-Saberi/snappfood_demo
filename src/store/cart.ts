import { createSlice } from "@reduxjs/toolkit";

interface Cart {
  productId: number;
  count: number;
}

const initialCartState: Array<Cart> = [];

const cart = createSlice({
  name: "cart",
  initialState: { cartUser: initialCartState },
  reducers: {
    emptyCart(state) {
      state.cartUser = [];
    },
    addToCart(state, action) {
      if (state.cartUser.some((item) => item.productId === action.payload)) {
        state.cartUser.map((item) => {
          if (item.productId === action.payload) {
            item.count++;
          }
        });
      } else {
        state.cartUser.push({ productId: action.payload, count: 1 });
      }
    },
    removeFromCart(state, action) {
      state.cartUser.map((item) => {
        if (item.productId === action.payload) {
          if (item.count === 1) {
            state.cartUser = state.cartUser.filter(
              (item) => item.productId !== action.payload
            );
          } else {
            item.count--;
          }
        }
      });
    },
  },
});

export const cartActions = cart.actions;
export default cart.reducer;
