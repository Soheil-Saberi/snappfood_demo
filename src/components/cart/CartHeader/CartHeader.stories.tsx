import { ComponentMeta, ComponentStory } from '@storybook/react';
import { Provider } from 'react-redux';
import store from '../../../store';
import CartHeader from './CartHeader';
import Cart from '../../../model/Cart';

interface Args {
  cartUser: Cart[];
}

export default {
  component: CartHeader,
  title: 'Cart/Header'
} as ComponentMeta<typeof CartHeader>;

export const CartHeaderStory: ComponentStory<typeof CartHeader> = (args: Args) => {
  return (
    <Provider store={store}>
      <div className="mt-4 gap-2 p-4 flex flex-col  border rounded-md shadow-md w-full h-fit">
        <CartHeader cartUser={args.cartUser} />
      </div>
    </Provider>
  );
};

CartHeaderStory.args = {
  cartUser: [
    { productId: 1, count: 1 },
    { productId: 2, count: 2 }
  ]
};
