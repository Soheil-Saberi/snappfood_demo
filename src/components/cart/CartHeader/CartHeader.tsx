import { Fragment, FC, useMemo } from 'react';
// Redux
import { useDispatch } from 'react-redux';
import { AppDispatch } from '../../../store';
import { cartActions } from '../../../store/cartSlice';
// Style
import styles from './CartHeader.module.css';
// Icons
import deleteIcon from '../../../assets/images/delete.png';
// Utils
import { sumCountCartUser, toFarsi } from '../../../utility/Utils';
// Model
import Cart from '../../../model/Cart';

interface Props {
  cartUser: Cart[];
}

const CartHeader: FC<Props> = ({ cartUser }) => {
  const dispatch = useDispatch<AppDispatch>();

  const memoSumOrderUser = useMemo(() => sumCountCartUser(cartUser), [cartUser]);

  return (
    <Fragment>
      {memoSumOrderUser !== '0' && (
        <div className={styles.div}>
          <p>سبد خرید ({toFarsi(memoSumOrderUser)})</p>
          <img onClick={() => dispatch(cartActions.emptyCart())} src={deleteIcon} alt="delete-icon" />
        </div>
      )}
    </Fragment>
  );
};

export default CartHeader;
