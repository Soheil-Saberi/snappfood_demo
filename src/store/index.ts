import { configureStore } from '@reduxjs/toolkit';

import cart from './cartSlice';

const store = configureStore({
  reducer: cart
});

export default store;
export type RootState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch;
