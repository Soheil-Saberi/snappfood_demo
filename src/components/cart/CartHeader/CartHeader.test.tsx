import renderer from 'react-test-renderer';
import { describe, expect, test } from '@jest/globals';
import { cleanup } from '@testing-library/react';
// Component
import CartHeader from './CartHeader';
// Model
import Cart from '../../../model/Cart';
// Redux
import { Provider } from 'react-redux';
import store from './../../../store/index';

afterEach(() => {
  cleanup();
});

describe('CartHeader Component', () => {
  test('Component Rendernig With Value Template (count = 1)', () => {
    const defaultValue: Cart[] = [{ productId: 1, count: 1 }];
    const cartHeader = renderer.create(
      <Provider store={store}>
        <CartHeader cartUser={defaultValue} />
      </Provider>
    );
    expect(cartHeader.toJSON()).toMatchSnapshot();
  });

  test('Component Rendernig With Empty Cart', () => {
    const defaultValue: Cart[] = [];
    const cartHeader = renderer.create(
      <Provider store={store}>
        <CartHeader cartUser={defaultValue} />
      </Provider>
    );
    expect(cartHeader.toJSON()).toBeNull();
  });
});
