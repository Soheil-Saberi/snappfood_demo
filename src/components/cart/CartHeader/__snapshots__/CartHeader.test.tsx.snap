// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CartHeader Component Component Rendernig With Value Template (count = 1) 1`] = `
<div>
  <p>
    سبد خرید (
    ۱
    )
  </p>
  <img
    alt="delete-icon"
    onClick={[Function]}
    src=""
  />
</div>
`;
